File: test/framework/mesh/mat_ids_from_function.lua
-- Setup mesh
nodes={}
N=10
L=5
xmin = -L/2
dx = L/N
for i=1,(N+1) do
    k=i-1
    nodes[i] = xmin + k*dx
end

meshgen1 = mesh.OrthogonalMeshGenerator.Create({ node_sets = {nodes,nodes,nodes} })
mesh.MeshGenerator.Execute(meshgen1)

--Sets a middle square to material 1
function mat_id(x,y,z,cur_id)
    if (math.abs(x)<L/10 and math.abs(y)<L/10) then
        return 1
    end
    return cur_id
end

mesh.SetMaterialIDFromFunction("mat_id")

mesh.ExportToVTK("new_mat_ids")

File: test/python/framework/mesh/mat_ids_from_function.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import sys
import math

if "opensn_console" not in globals():
    from mpi4py import MPI
    size = MPI.COMM_WORLD.size
    rank = MPI.COMM_WORLD.rank
    sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../../")))
    from pyopensn.mesh import OrthogonalMeshGenerator, KBAGraphPartitioner
    from pyopensn.xs import MultiGroupXS
    from pyopensn.source import VolumetricSource
    from pyopensn.aquad import GLProductQuadrature1DSlab
    from pyopensn.solver import DiscreteOrdinatesSolver, SteadyStateSolver
    from pyopensn.fieldfunc import FieldFunctionGridBased
    from pyopensn.fieldfunc import FieldFunctionInterpolationLine, FieldFunctionInterpolationVolume
    from pyopensn.settings import EnableCaliper
    from pyopensn.math import Vector3
    from pyopensn.logvol import RPPLogicalVolume

if __name__ == "__main__":

    # Setup mesh
    nodes = []
    N = 10
    L = 5
    xmin = -L / 2
    dx = L / N
    for i in range(N + 1):
        nodes.append(xmin + i * dx)
    meshgen = OrthogonalMeshGenerator(node_sets=[nodes, nodes, nodes])
    grid = meshgen.Execute()

    # Sets a middle square to material 1
    def mat_id(pt, cur_id):
        if abs(pt.x) < L / 10 and abs(pt.y) < L / 10:
            return 1
        return cur_id

    grid.SetBlockIDFromFunction(mat_id)
    grid.ExportToPVTU("new_mat_ids")

    if rank == 0:
        os.system("rm new_mat_ids.pvtu")
        os.system("rm new_mat_ids_0.vtu")
